service/history/wrappers/ratelimited/handler_generated.go

```diff
@@ -0,0 +1,314 @@
+}
+
+func (h *historyHandler) RespondActivityTaskCanceled(ctx context.Context, hp1 *types.HistoryRespondActivityTaskCanceledRequest) (err error) {
+	return h.wrapped.RespondActivityTaskCanceled(ctx, hp1)
+}
+
+func (h *historyHandler) RespondActivityTaskCompleted(ctx context.Context, hp1 *types.HistoryRespondActivityTaskCompletedRequest) (err error) {
+	return h.wrapped.RespondActivityTaskCompleted(ctx, hp1)
+}
+
+func (h *historyHandler) RespondActivityTaskFailed(ctx context.Context, hp1 *types.HistoryRespondActivityTaskFailedRequest) (err error) {
+	return h.wrapped.RespondActivityTaskFailed(ctx, hp1)
+}
+
+func (h *historyHandler) RespondCrossClusterTasksCompleted(ctx context.Context, rp1 *types.RespondCrossClusterTasksCompletedRequest) (rp2 *types.RespondCrossClusterTasksCompletedResponse, err error) {
+	return h.wrapped.RespondCrossClusterTasksCompleted(ctx, rp1)
+}
+
+func (h *historyHandler) RespondDecisionTaskCompleted(ctx context.Context, hp1 *types.HistoryRespondDecisionTaskCompletedRequest) (hp2 *types.HistoryRespondDecisionTaskCompletedResponse, err error) {
+	return h.wrapped.RespondDecisionTaskCompleted(ctx, hp1)
+}
+
+func (h *historyHandler) RespondDecisionTaskFailed(ctx context.Context, hp1 *types.HistoryRespondDecisionTaskFailedRequest) (err error) {
+	return h.wrapped.RespondDecisionTaskFailed(ctx, hp1)
+}
+
+func (h *historyHandler) ScheduleDecisionTask(ctx context.Context, sp1 *types.ScheduleDecisionTaskRequest) (err error) {
+	return h.wrapped.ScheduleDecisionTask(ctx, sp1)
+}
+
+func (h *historyHandler) SignalWithStartWorkflowExecution(ctx context.Context, hp1 *types.HistorySignalWithStartWorkflowExecutionRequest) (sp1 *types.StartWorkflowExecutionResponse, err error) {
+
+	if hp1 == nil {
+		err = validate.ErrRequestNotSet
+		return
+	}
+
+	if hp1.GetDomainUUID() == "" {
+		err = validate.ErrDomainNotSet
+		return
+	}
+
+	if hp1.SignalWithStartRequest.GetWorkflowID() == "" {
+		err = validate.ErrWorkflowIDNotSet
+		return
+	}
+
+	h.workflowIDCache.AllowExternal(hp1.SignalWithStartRequest.GetWorkflowID(), hp1.GetDomainUUID())
+	return h.wrapped.SignalWithStartWorkflowExecution(ctx, hp1)
+}
+
+func (h *historyHandler) SignalWorkflowExecution(ctx context.Context, hp1 *types.HistorySignalWorkflowExecutionRequest) (err error) {
+
+	if hp1 == nil {
+		err = validate.ErrRequestNotSet
+		return
+	}
+
+	if hp1.GetDomainUUID() == "" {
+		err = validate.ErrDomainNotSet
+		return
+	}
+
+	if hp1.SignalRequest.GetWorkflowExecution().GetWorkflowID() == "" {
+		err = validate.ErrWorkflowIDNotSet
+		return
+	}
+
+	h.workflowIDCache.AllowExternal(hp1.SignalRequest.GetWorkflowExecution().GetWorkflowID(), hp1.GetDomainUUID())
+	return h.wrapped.SignalWorkflowExecution(ctx, hp1)
+}
+
+func (h *historyHandler) Start() {
+	h.wrapped.Start()
+	return
+}
+
+func (h *historyHandler) StartWorkflowExecution(ctx context.Context, hp1 *types.HistoryStartWorkflowExecutionRequest) (sp1 *types.StartWorkflowExecutionResponse, err error) {
```

Comment:
This approach would work fine but we will introduce other rate limiting types/signals in the near future. If each different rate limit type requires generating a new layer of handler it will be too much extra code/layers. Did you consider enhancing `rateLimiter` inside handler.go so it can support chained limiters including workflowid one?

Reply:
Hi, I did it like this because we have just introduced wrappers for rate limiters in the frontend. I think making a generalised chain-able rate limiting architecture, where we can target the different methods with different rate limiters, is probably too much scope for the current project. 

I'm not a huge fan of the wrappers TBH. An alternative would be to just write the straightforward code in the methods. IMO that would be much more readable, and easier to discover.

Reply:
Yeah there's no single best approach. I think sticking with the similar approach as other services (e.g. frontend) do this at least makes it consistent so that makes sense. However, the frontend handler object does not have the `rateLimiter` itself. It delegates that responsibility to [generated wrapper](https://github.com/uber/cadence/blob/master/service/frontend/wrappers/ratelimited/ratelimit.go#L38-L39). 
In this PR, we have workflowid rate limiter as a wrapper but there's still existing rate limiter in history handler. That's why I was thinking one wrapper per rate limiter type is too much code to generate/navigate. If you move the existing history handler's rate limiter to the new wrapper then it would be more consistent. 

Also @Groxx you have been more focused on rate limiters lately and probably have an opinion here. How were you planning to introduce the globabal rate limiter in service layer?

Reply:
Hi, I think we should absolutely move the history host handler to this wrapper, so all rate limiting is collected here, I'd suggest to do that in the next PR?

Reply:
I think we can pull it out in a function, like in front end. Could be easier to test.

Not a huge fan of passing the request type to another function, but it's OK, then we have alignment across the services

Reply:
Makes total sense to do that part in follow up PR.